pkg/policy/api/ingress.go

```diff
@@ -237,3 +252,60 @@ func (i *IngressCommonRule) GetSourceEndpointSelectorsWithRequirements(requireme
 func (i *IngressCommonRule) AllowsWildcarding() bool {
 	return len(i.FromRequires) == 0
 }
+
+// RequiresDerivative returns true when the EgressCommonRule contains sections
+// that need a derivative policy created in order to be enforced
+// (e.g. ToGroups).
+func (e *IngressCommonRule) RequiresDerivative() bool {
+	return len(e.FromGroups) > 0
+}
+
+// CreateDerivative will return a new rule based on the data gathered by the
+// rules that creates a new derivative policy.
+// In the case of FromGroups will call outside using the groups callback and this
+// function can take a bit of time.
+func (e *IngressRule) CreateDerivative(ctx context.Context) (*IngressRule, error) {
+	newRule := e.DeepCopy()
+	if !e.RequiresDerivative() {
+		return newRule, nil
+	}
+	newRule.FromCIDRSet = make(CIDRRuleSlice, 0, len(e.FromGroups))
+	for _, group := range e.FromGroups {
+		cidrSet, err := group.GetCidrSet(ctx)
+		if err != nil {
+			return &IngressRule{}, err
+		}
+		if len(cidrSet) == 0 {
+			return &IngressRule{}, nil
+		}
+		newRule.FromCIDRSet = append(e.FromCIDRSet, cidrSet...)
+	}
+	newRule.FromGroups = nil
+	e.SetAggregatedSelectors()
+	return newRule, nil
+}
+
+// CreateDerivative will return a new rule based on the data gathered by the
+// rules that creates a new derivative policy.
+// In the case of FromGroups will call outside using the groups callback and this
+// function can take a bit of time.
+func (e *IngressDenyRule) CreateDerivative(ctx context.Context) (*IngressDenyRule, error) {
+	newRule := e.DeepCopy()
+	if !e.RequiresDerivative() {
+		return newRule, nil
+	}
+	newRule.FromCIDRSet = make(CIDRRuleSlice, 0, len(e.FromGroups))
+	for _, group := range e.FromGroups {
+		cidrSet, err := group.GetCidrSet(ctx)
+		if err != nil {
+			return &IngressDenyRule{}, err
+		}
+		if len(cidrSet) == 0 {
+			return &IngressDenyRule{}, nil
+		}
+		newRule.FromCIDRSet = append(e.FromCIDRSet, cidrSet...)
+	}
+	newRule.FromGroups = nil
+	e.SetAggregatedSelectors()
+	return newRule, nil
+}
```

Comment:
I'd suggest looking to see if we can reasonably DRY up this code from its egress counterpart.

Reply:
Hey Chris, I'll tidy everything else up but looking at this code I'm not sure it really makes a huge amount of sense to make it more DRY. I know it's repetitive but most of the logic is calls to struct methods, so abstracting this out was just getting ridiculously complicated.

Reply:
Could we at least have a common utility function that accepts the `api.Groups` type and returns the slice of CIDR sets? Basically, it could abstract away the for-loop above.